core/src/com/unciv/logic/battle/GreatGeneralImplementation.kt

```diff
@@ -0,0 +1,141 @@
+     */
+    fun getGreatGeneralBonus(unit: MapUnit): Int {
+        val civInfo = unit.civInfo
+        val nearbyGenerals = unit.getTile()
+            .getTilesInDistance(civInfo.maxGeneralBonusRadius)
+            .flatMap { it.getUnits() }
+            .filter { it.civInfo == civInfo && it.hasGreatGeneralUnique }
+        if (nearbyGenerals.none()) return 0
+
+        val greatGeneralModifier = nearbyGenerals
+            .flatMap { general ->
+                general.getMatchingUniques(UniqueType.GreatGeneralAura)
+                    .map { GeneralBonusData(general, it) } +
+                        general.getMatchingUniques(UniqueType.BonusForUnitsInRadius)
+                            .map { GeneralBonusData(general) }
+            }.filter {
+                // Support the border case when a mod unit has several
+                // GreatGeneralAura uniques (e.g. +50% as radius 1, +25% at radius 2, +5% at radius 3)
+                it.general.currentTile.aerialDistanceTo(unit.getTile()) <= it.radius
+                        && (it.filter == "All" || unit.matchesFilter(it.filter))
+            }.maxOfOrNull { it.bonus } ?: 0
+        if (unit.hasUnique(UniqueType.GreatGeneralProvidesDoubleCombatBonus, checkCivInfoUniques = true))
+            return greatGeneralModifier * 2
+        return greatGeneralModifier
+    }
+
+    /**
+     * Find a tile for accompanying a military unit where the total bonus for all affected units is maximized.
+     *
+     * Used by [SpecificUnitAutomation.automateGreatGeneral].
+     */
+    fun getBestAffectedTroopsTile(general: MapUnit): TileInfo? {
+        // Normally we have only one Unique here. But a mix is not forbidden, so let's try to support mad modders.
+        // (imagine several GreatGeneralAura uniques - +50% at radius 1, +25% at radius 2, +5% at radius 3 - possibly learnable from promotions via buildings or natural wonders?)
+
+        // Map out the uniques sorted by bonus, as later only the best bonus will apply.
+        val generalBonusData = (
+                general.getMatchingUniques(UniqueType.GreatGeneralAura).map { GeneralBonusData(general, it) } +
+                general.getMatchingUniques(UniqueType.BonusForUnitsInRadius).map { GeneralBonusData(general) }
+            ).sortedWith(compareByDescending<GeneralBonusData> { it.bonus }.thenBy { it.radius })
+            .toList()
+
+        // Get candidate units to 'follow', coarsely.
+        // The mapUnitFilter of the unique won't apply here but in the ranking of the "Aura" effectiveness.
+        val unitMaxMovement = general.getMaxMovement()
+        val militaryUnitTilesInDistance = general.movement.getDistanceToTiles().asSequence()
+            .map { it.key }
+            .filter { tile ->
+                val militaryUnit = tile.militaryUnit
+                militaryUnit != null && militaryUnit.civInfo == general.civInfo
+                        && (tile.civilianUnit == null || tile.civilianUnit == general)
```

Comment:
What if the unit with such "Strength bonus in radius" unique will be not a civilian (as we got used for) but a military unit?

Reply:
I briefly thought about it, and am convinced such mods already exist. I dismissed the train of thought as battle damage will work, but automation wouldn't, as before. That isCivilian check early on... There's also that "AI wouldn't know what to do with your Archeologist" check in modchecker, and solving that would maybe become easier with a quite different structure - a little fuzzy, actually evaluating more alternatives and ranking them, that direction - not in this PR.

Until then, we have a guarantee that this code will only see civilians.

Even if - the check is to see whether the tile has "room" for the "general" assuming a general is always a civilian, and if it were not, it would still head toward that tile but stop short, a feature of UnitMovementAlgorithms.